cpp/src/parquet/util/crypto.h

```diff
@@ -18,29 +18,126 @@
 #ifndef PARQUET_UTIL_CRYPTO_H
 #define PARQUET_UTIL_CRYPTO_H
 
+#include <list>
 #include <memory>
+#include <string>
 
+#include <openssl/evp.h>
 #include "parquet/properties.h"
 #include "parquet/types.h"
 
-using parquet::Encryption;
-using parquet::EncryptionProperties;
+using parquet::ParquetCipher;
 
 namespace parquet_encryption {
 
-int Encrypt(Encryption::type alg_id, bool metadata, const uint8_t* plaintext,
-            int plaintext_len, uint8_t* key, int key_len, uint8_t* aad, int aad_len,
-            uint8_t* ciphertext);
+constexpr int GCMTagLength = 16;
+constexpr int NonceLength = 12;
 
-int Encrypt(std::shared_ptr<EncryptionProperties> encryption_props, bool metadata,
-            const uint8_t* plaintext, int plaintext_len, uint8_t* ciphertext);
+// Module types
+const int8_t Footer = 0;
+const int8_t ColumnMetaData = 1;
+const int8_t DataPage = 2;
+const int8_t DictionaryPage = 3;
+const int8_t DataPageHeader = 4;
+const int8_t DictionaryPageHeader = 5;
+const int8_t ColumnIndex = 6;
+const int8_t OffsetIndex = 7;
 
-int Decrypt(Encryption::type alg_id, bool metadata, const uint8_t* ciphertext,
-            int ciphertext_len, uint8_t* key, int key_len, uint8_t* aad, int aad_len,
-            uint8_t* plaintext);
+class AesEncryptor {
+ public:
+  // Can serve one key length only. Possible values: 16, 24, 32 bytes.
+  AesEncryptor(ParquetCipher::type alg_id, int key_len, bool metadata,
+               std::shared_ptr<std::list<AesEncryptor*>> all_encryptors);
+
+  // Size different between plaintext and ciphertext, for this cipher.
+  int CiphertextSizeDelta();
+
+  // Key length is passed only for validation. If different from value in
+  // constructor, exception will be thrown.
+  int Encrypt(const uint8_t* plaintext, int plaintext_len, uint8_t* key, int key_len,
+              uint8_t* aad, int aad_len, uint8_t* ciphertext);
+
+  int SignedFooterEncrypt(const uint8_t* footer, int footer_len, uint8_t* key,
+                          int key_len, uint8_t* aad, int aad_len, uint8_t* nonce,
+                          uint8_t* encrypted_footer);
+
+  void WipeOut() {
+    if (NULLPTR != ctx_) {
+      EVP_CIPHER_CTX_free(ctx_);
+      ctx_ = NULLPTR;
+    }
+  }
+
+  ~AesEncryptor() {
+    if (NULLPTR != ctx_) {
+      EVP_CIPHER_CTX_free(ctx_);
+      ctx_ = NULLPTR;
+    }
+  }
+
+ private:
+  EVP_CIPHER_CTX* ctx_;
```

Comment:
Can you hide this in a PIMPL? If this is an internal header then this is not necessary